src/services/api/activity-log/activity.log.builder.service.ts

```diff
@@ -0,0 +1,165 @@
+import { IActivity } from '@platform/activity';
+import { ActivityEventType } from '@common/enums/activity.event.type';
+import { IActivityLogEntryMemberJoined } from '@services/api/activity-log/dto/activity.log.dto.entry.member.joined.interface';
+import { IActivityLogEntryCalloutPublished } from '@services/api/activity-log/dto/activity.log.dto.entry.callout.published';
+import { IActivityLogEntryCalloutCardCreated } from '@services/api/activity-log/dto/activity.log.dto.entry.callout.card.created';
+import { IActivityLogEntryCalloutCanvasCreated } from '@services/api/activity-log/dto/activity.log.dto.entry.callout.canvas.created';
+import { IActivityLogEntryChallengeCreated } from '@services/api/activity-log/dto/activity.log.dto.entry.challenge.created';
+import { IActivityLogEntryOpportunityCreated } from '@services/api/activity-log/dto/activity.log.dto.entry.opportunity.created';
+import { IActivityLogEntryCalloutCardComment } from '@services/api/activity-log/dto/activity.log.dto.entry.callout.card.comment';
+import { IActivityLogEntryCalloutDiscussionComment } from '@services/api/activity-log/dto/activity.log.dto.entry.callout.discussion.comment';
+import { IUser } from '@src/domain';
+import { ActivityService } from '@platform/activity/activity.service';
+import { UserService } from '@domain/community/user/user.service';
+import { CalloutService } from '@domain/collaboration/callout/callout.service';
+import { AspectService } from '@domain/collaboration/aspect/aspect.service';
+import { CanvasService } from '@domain/common/canvas/canvas.service';
+import { ChallengeService } from '@domain/challenge/challenge/challenge.service';
+import { OpportunityService } from '@domain/collaboration/opportunity/opportunity.service';
+import { CommunityService } from '@domain/community/community/community.service';
+
+interface ActivityLogBuilderFunction<TypedActivityLogEntry> {
+  (rawActivity: IActivity): Promise<TypedActivityLogEntry>;
+}
+
+interface IActivityLogBuilder {
+  [ActivityEventType.MEMBER_JOINED]: ActivityLogBuilderFunction<IActivityLogEntryMemberJoined>;
+  [ActivityEventType.CALLOUT_PUBLISHED]: ActivityLogBuilderFunction<IActivityLogEntryCalloutPublished>;
+  [ActivityEventType.CARD_CREATED]: ActivityLogBuilderFunction<IActivityLogEntryCalloutCardCreated>;
+  [ActivityEventType.CANVAS_CREATED]: ActivityLogBuilderFunction<IActivityLogEntryCalloutCanvasCreated>;
+  [ActivityEventType.CHALLENGE_CREATED]: ActivityLogBuilderFunction<IActivityLogEntryChallengeCreated>;
+  [ActivityEventType.OPPORTUNITY_CREATED]: ActivityLogBuilderFunction<IActivityLogEntryOpportunityCreated>;
+  [ActivityEventType.CARD_COMMENT]: ActivityLogBuilderFunction<IActivityLogEntryCalloutCardComment>;
+  [ActivityEventType.DISCUSSION_COMMENT]: ActivityLogBuilderFunction<IActivityLogEntryCalloutDiscussionComment>;
+}
+
+export interface ActivityLogEntryBase {
```

Comment:
One type per file. That is what is done everywhere else on the server and lets keep to that pattern unless we explicitly decide to move away from this. So please create an activity.log.entry.base.interface file with this type in it. Also type name should start with "I" to match the naming convention on the server.